File: problem/Remove_Duplicates_from_Sorted_Array/removeDuplicates.c
int removeDuplicates(int *nums, int numsSize)
{
    if (numsSize == 0)
        return 0;

    int slow = 0;
    for (int fast = 1; fast < numsSize; fast++)
    {
        if (nums[fast] != nums[slow])
        {
            slow++;
            nums[slow] = nums[fast];
        }
    }

    return slow + 1;
}

int removeDuplicates_alternative(int *nums, int numsSize)
{
    if (numsSize == 0)
        return 0;

    int writeIndex = 1;
    for (int i = 1; i < numsSize; i++)
    {
        if (nums[i] != nums[i - 1])
        {
            nums[writeIndex] = nums[i];
            writeIndex++;
        }
    }

    return writeIndex;
}

int removeDuplicates_verbose(int *nums, int numsSize)
{
    if (numsSize <= 1)
        return numsSize;

    int uniqueCount = 1;
    for (int i = 1; i < numsSize; i++)
    {
        if (nums[i] != nums[uniqueCount - 1])
        {
            nums[uniqueCount] = nums[i];
            uniqueCount++;
        }
    }

    return uniqueCount;
}

File: problem/Remove_Duplicates_from_Sorted_Array/removeDuplicates.js
/**
 * Two-pointer approach - O(n) time, O(1) space
 */
var removeDuplicates = function(nums) {
    if (nums.length === 0)
        return 0;
    
    let slow = 0;
    for (let fast = 1; fast < nums.length; fast++) {
        if (nums[fast] !== nums[slow]) {
            slow++;
            nums[slow] = nums[fast];
        }
    }
    
    return slow + 1;
};

/**
 * Alternative implementation
 */
var removeDuplicates_alternative = function(nums) {
    if (nums.length === 0) 
        return 0;
    
    let writeIndex = 1;
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] !== nums[i - 1]) {
            nums[writeIndex] = nums[i];
            writeIndex++;
        }
    }
    
    return writeIndex;
};

/**
 * Using Set for comparison (not in-place, for reference)
 */
var removeDuplicates_set = function(nums) {
    const uniqueElements = [...new Set(nums)];
    
    for (let i = 0; i < uniqueElements.length; i++) {
        nums[i] = uniqueElements[i];
    }
    
    return uniqueElements.length;
};
